### PR TITLE
Ac matching

### PR DESCRIPTION
**Pull request by nwatx**

### CODE REVIEW

**Comment by bqi343** [comment]
@thecodingwizard what's going on w/ deployment?

File: solutions/AC Matching.mdx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+id: ac-matching
+source: AtCoder DP Contest
+title: Matching
+author: Neo Wang
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
+$S$, this problem boils down to the following formulation:
```
Comment (bqi343) [inline solutions/AC Matching.mdx:11]: Seems pretty difficult to understand. Shouldn't $dp[S]$ be the number of ways to match the females in $S$ to the first $|S|$ males?

File: solutions/AC Matching.mdx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#include <bits/stdc++.h>
+using namespace std;
+
+bool compat[21][21];
+int dp[1 << 21];
+
+int N;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
```
Comment (bqi343) [inline solutions/AC Matching.mdx:40]: messed up indentation

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Comment by thecodingwizard** [comment]
> @thecodingwizard what's going on w/ deployment?

Oops I accidentally broke a dependency (fixed on master). Either merge master into this branch or just merge this branch into master and trust that build will work 😅

File: solutions/AC Matching.mdx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+author: Neo Wang
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
+$S$ to the first $|S|$ males, this problem boils down to the following
+formulation:
+
+$$
```
Comment (bqi343) [inline solutions/AC Matching.mdx:14]: what's \text{male}

File: solutions/AC Matching.mdx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[ \text{male}][x]
+$$
+
+In English, this is the equivalent of
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
```
Comment (bqi343) [inline solutions/AC Matching.mdx:20]: what does "produces a match" mean?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/AC Matching.mdx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+int N;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	cin >> N;
+	for(int i = 0; i < N; i++) for(int j = 0; j < N; j++) cin >> compat[i][j];
+
+	for(int i = 0; i < N; i++) if(compat[0][i]) dp[1 << i] = 1;
+	for(int s = 1; s < (1<<N); s++) {
```
Comment (bqi343) [inline solutions/AC Matching.mdx:46]: Can't you set `dp[0]=1` and start the loop at `s=0`?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/AC Matching.mdx (hunk 5 above)
Comment (nwatx) [reply solutions/AC Matching.mdx]: yeah that works too.

File: solutions/AC Matching.mdx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
+$$
+
+In English, this is the equivalent of
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
+> equivalent to the sum of all the matchings without female $x$ such that the
+> addition of female $x$ is compatible with the $|S|$-th male.
+
+Our base cases are built using matches of the first male; if the first male is
```
Comment (bqi343) [inline solutions/AC Matching.mdx:24]: update this

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/AC Matching.mdx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
+$$
+
+In English, this is the equivalent of
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
+> equivalent to the sum of all the matchings without female $x$ such that the
+> addition of female $x$ is compatible with the $|S|$-th male.
+
+Our base case is the empty set, which has a value of $1$. Although the empty set
```
Comment (bqi343) [inline solutions/AC Matching.mdx:24]: > Although the empty set itself contains zero matchings

why wouldn't the empty set be considered to generate a single (empty) matching?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/AC Matching.mdx (hunk 7 above)
Comment (nwatx) [reply solutions/AC Matching.mdx:24]: isn't dp[s] the amount of matchings for females in set|s| to the first s males? If so then wouldn't the empty set contain zero matchings as there are no females. Also doesn't "generating the base cases" indicate that the empty set is used to generate a single matching.

Comment (nwatx) [reply solutions/AC Matching.mdx:24]: aka "it generates a matching, but by itself does not contain any match" which is valid because N>=1"

Comment (bqi343) [reply solutions/AC Matching.mdx:24]: > isn't dp[s] the amount of matchings for females in set|s| to the first s males?

yes

> If so then wouldn't the empty set contain zero matchings as there are no females

Er, I would consider the empty set as a single matching involving zero pairs.

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/AC Matching.mdx (hunk 7 above)
Comment (nwatx) [reply solutions/AC Matching.mdx]: oh that's pretty neat.

File: solutions/AC Matching.mdx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	cin.tie(0)->sync_with_stdio(0);
+	cin >> N;
+	for(int i = 0; i < N; i++) for(int j = 0; j < N; j++) cin >> compat[i][j];
+
+	dp[0] = 1;
+
+	for(int s = 0; s < (1<<N); s++) {
+		int v = __builtin_popcount(s);
+		for(int x = 0; x < N; x++) {
+			if((s & (1 << x)) || !compat[v][x]) continue;
```
Comment (bqi343) [inline solutions/AC Matching.mdx:50]: `(s & (1 << x))`

why is there space for bitwise and but not bitwise or? Same question regarding the bit shifts above and below

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by bqi343 — APPROVED** [review]

File: solutions/AC Matching.mdx [hunk 9]
```diff
@@ -0,0 +1,58 @@
+---
+id: ac-matching
+source: AtCoder DP Contest
+title: Matching
+author: Neo Wang
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
```
Comment (dolphingarlic) [inline solutions/AC Matching.mdx:10]:
```suggestion
If we define $\texttt{dp}[S]$ to be the number of matchings for females in the set
```

File: solutions/AC Matching.mdx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+$S$ to the first $|S|$ males, this problem boils down to the following
+formulation:
+
+$$
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
+$$
+
+In English, this is the equivalent of
```
Comment (dolphingarlic) [inline solutions/AC Matching.mdx:18]:
```suggestion
In English, this is equivalent to the following:
```

File: solutions/AC Matching.mdx [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
```
Comment (dolphingarlic) [inline solutions/AC Matching.mdx:20]:
```suggestion
> The number of matchings in a subset $S$ to include a certain female $x$ is
```

File: solutions/AC Matching.mdx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
+$S$ to the first $|S|$ males, this problem boils down to the following
+formulation:
+
+$$
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
```
Comment (dolphingarlic) [inline solutions/AC Matching.mdx:15]: Instead of using \text{if}, use a single colon

File: solutions/AC Matching.mdx [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
+> equivalent to the sum of all the matchings without female $x$ such that the
+> addition of female $x$ is compatible with the $|S|$-th male.
+
+Our base case is the empty set, which has a value of $1$ (the empty set can be
+considered as a single matching involving zero pairs).
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\cdot 2^N)$
```
Comment (dolphingarlic) [inline solutions/AC Matching.mdx:29]: Move the time complexity to the top

File: solutions/AC Matching.mdx [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+$$
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
+$$
+
+In English, this is the equivalent of
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
+> equivalent to the sum of all the matchings without female $x$ such that the
+> addition of female $x$ is compatible with the $|S|$-th male.
```
Comment (dolphingarlic) [inline solutions/AC Matching.mdx:22]: > the addition of female $x$ is compatible with the $|S|$-th male.

Do you mean that female $x$ is compatible or that the addition is compatible?

File: solutions/AC Matching.mdx [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+source: AtCoder DP Contest
+title: Matching
+author: Neo Wang
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
+$S$ to the first $|S|$ males, this problem boils down to the following
+formulation:
```
Comment (dolphingarlic) [inline solutions/AC Matching.mdx:12]:
```suggestion
$S$ to the first $|S|$ males, this problem boils down to the following:
```

**Review by dolphingarlic — CHANGES_REQUESTED** [review]

**Review by dongliu0426 — APPROVED** [review]

**Pull request merged**